tests/unittests/TraceEventsTest.cpp

```diff
@@ -567,7 +567,78 @@ TEST_P(TraceEventsTest, deviceManagerEvents) {
   auto &traceEvents = context.getTraceContext()->getTraceEvents();
 
   ASSERT_GT(traceEvents.size(), 0);
-  checkEventTimestamps(traceEvents);
+  // CompleteEvents are not necessarily monitonically increasing since they are
+  // added to the log when they end, not when they start.
+}
+
+TEST(TraceEventsTest, nestedScopedEvents) {
+  ExecutionContext context;
+  context.setTraceContext(
+      llvm::make_unique<TraceContext>(TraceLevel::STANDARD));
+
+  TraceContext *tc = context.getTraceContext();
+
+  ScopedTraceBlock block_one(tc, "one");
+  {
+    ScopedTraceBlock block_two(tc, "two");
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+  }
+
+  {
+    ScopedTraceBlock block_three(tc, "three");
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    {
+      ScopedTraceBlock block_four(tc, "four");
+      std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    }
+  }
+
+  block_one.end();
+
+  auto &traceEvents = tc->getTraceEvents();
+  ASSERT_EQ(traceEvents.size(), 4);
+  llvm::StringMap<uint64_t> durations;
+  for (auto &tc : traceEvents) {
+    durations[tc.name] = tc.duration;
+  }
+
+  ASSERT_GE(durations["one"], durations["two"] + durations["three"]);
+  ASSERT_GE(durations["three"], durations["four"]);
+}
+
+TEST(TraceEventsTest, nestedScopedEventsMacro) {
+  ExecutionContext context;
+  context.setTraceContext(
+      llvm::make_unique<TraceContext>(TraceLevel::STANDARD));
+
+  TraceContext *tc = context.getTraceContext();
+
+  TRACE_EVENT_SCOPE(tc, "one");
+  {
+    TRACE_EVENT_SCOPE_NAMED(tc, "two", block_two);
```

Comment:
Nit: do `block_two`, `block_three` and `block_four` have to be named scopes?

Reply:
Nope, I'll move them to be unnamed and test named scopes in the next test